File: implementations/typescript/src/harness.js
import { validateEnvelope } from "./validate.js";
import { isValidBySchema } from "./schema.js";
import { AepSession } from "./session.js";
import { TaskTracker } from "./task.js";
import { EventRouter } from "./router.js";
import { ErrorCode, errorPayload } from "./errors.js";
import { isStandardEventType } from "./event-types.js";
import { DeliveryTracker } from "./delivery.js";

export class AepHarness {
  constructor(options = {}) {
    this.source = options.source ?? "harness:aep";
    this.now = options.now ?? (() => new Date().toISOString());
    this.useSchemaValidation = options.useSchemaValidation ?? false;
    this._sequence = 0;
    this._subscriptions = new Map();
    this._tasks = new Map();
    this._router = new EventRouter();
    this._session = null;
    this.delivery = new DeliveryTracker(options.delivery);

    this._setupRouter();
    this._setupDeliveryRouter();
  }

  get session() { return this._session; }
  get subscriptions() { return new Map(this._subscriptions); }
  get tasks() { return new Map(this._tasks); }

  _setupRouter() {
    this._router
      .on("capabilities.requested", (event) => this._handleCapabilitiesRequested(event))
      .on("subscription.requested", (event) => this._handleSubscriptionRequested(event))
      .on("subscription.cancelled", (event) => this._handleSubscriptionCancelled(event))
      .on("task.submitted", (event) => this._handleTaskSubmitted(event))
      .on((event) => event.type.startsWith("task.") && event.type !== "task.submitted", (event) => this._handleTaskEvent(event))
      .on("session.opened", (event) => this._handleSessionOpened(event))
      .on("session.closed", (event) => this._handleSessionClosed(event));
  }

  _setupDeliveryRouter() {
    this._router
      .on("event.acknowledged", (event) => this._handleInboundAck(event))
      .on("event.redelivered", (event) => this._handleInboundRedeliver(event))
      .on("event.dead_lettered", (event) => this._handleInboundDeadLetter(event));
  }

  handle(value) {
    const errors = validateEnvelope(value);
    if (errors.length > 0) {
      return [this._event("event.rejected", value, {
        errors,
        error: errorPayload(ErrorCode.INVALID_ENVELOPE, errors[0])
      })];
    }

    if (this.useSchemaValidation && !isValidBySchema(value, "envelope")) {
      return [this._event("event.rejected", value, {
        errors: ["schema validation failed"],
        error: errorPayload(ErrorCode.INVALID_ENVELOPE, "envelope does not conform to JSON Schema")
      })];
    }

    if (!isStandardEventType(value.type) && !value.type.startsWith("session.")) {
      return [this._event("event.rejected", value, {
        errors: [`type not in standard draft registry: ${value.type}`],
        error: errorPayload(ErrorCode.INVALID_EVENT_TYPE, `unknown event type: ${value.type}`, { retryable: false })
      })];
    }

    if (value.spec_version !== "0.2") {
      return [this._event("event.rejected", value, {
        errors: [`unsupported protocol version: ${value.spec_version}`],
        error: errorPayload(ErrorCode.UNSUPPORTED_VERSION, `unsupported version ${value.spec_version}`, { details: { supported: ["0.2"] } })
      })];
    }

    if (value.delivery?.mode) {
      this.delivery.track(value.id, value.session_id ?? "_default");
    }

    const routed = this._router.dispatch(value);
    if (routed.length > 0) return routed;

    return [this._event("event.acknowledged", value, {
      acknowledged_event_id: value.id
    })];
  }

  _handleInboundAck(event) {
    const eventId = event.payload?.acknowledged_event_id;
    if (eventId) this.delivery.ack(eventId);
    return [this._event("event.acknowledged", event, {
      acknowledged_event_id: event.id
    })];
  }

  _handleInboundRedeliver(event) {
    const eventId = event.payload?.original_event_id;
    if (eventId) this.delivery.nack(eventId);
    return [this._event("event.acknowledged", event, {
      acknowledged_event_id: event.id
    })];
  }

  _handleInboundDeadLetter(event) {
    const eventId = event.payload?.original_event_id;
    if (eventId) {
      this.delivery.deadLetter(eventId, event.payload?.error ?? { code: "unknown" });
    }
    return [this._event("event.acknowledged", event, {
      acknowledged_event_id: event.id
    })];
  }

  _handleCapabilitiesRequested(event) {
    return this._event("capabilities.declared", event, {
      protocol: "harmovela",
      spec_version: "0.2",
      transports: ["stdio"],
      delivery_modes: ["best_effort", "at_least_once", "replayable"],
      features: [
        "envelope_validation",
        "event_type_registry",
        "subscription_matching",
        "session_lifecycle",
        "task_lifecycle",
        "error_model",
        "event_routing"
      ]
    });
  }

  _handleSubscriptionRequested(event) {
    const payload = event.payload ?? {};
    const subscriptionId = `sub_${String(++this._sequence).padStart(4, "0")}`;

    if (!payload.types && !payload.source && !payload.target && !payload.topic) {
      return this._event("subscription.rejected", event, {
        subscription_id: subscriptionId,
        filter: payload,
        error: errorPayload(ErrorCode.SUBSCRIPTION_REJECTED, "subscription must include at least one filter criterion")
      });
    }

    this._subscriptions.set(subscriptionId, {
      id: subscriptionId,
      filter: payload,
      created_at: this.now()
    });

    return this._event("subscription.created", event, {
      subscription_id: subscriptionId,
      filter: payload
    });
  }

  _handleSubscriptionCancelled(event) {
    const subId = event.payload?.subscription_id;
    if (subId && this._subscriptions.has(subId)) {
      this._subscriptions.delete(subId);
    }
    return this._event("event.acknowledged", event, {
      acknowledged_event_id: event.id
    });
  }

  _handleTaskSubmitted(event) {
    const taskId = event.task_id ?? event.payload?.task_id ?? `task_${Date.now().toString(36)}`;
    if (this._tasks.has(taskId)) {
      return this._event("event.rejected", event, {
        error: errorPayload(ErrorCode.TASK_ERROR, `duplicate task id: ${taskId}`)
      });
    }

    const tracker = new TaskTracker({
      task_id: taskId,
      description: event.payload?.description,
      source: event.source,
      session_id: event.session_id,
      conversation_id: event.conversation_id,
      created_at: event.created_at
    });

    this._tasks.set(taskId, tracker);
    tracker.accepted();

    return this._event("task.accepted", event, {
      task_id: taskId,
      status: "accepted"
    });
  }

  _handleTaskEvent(event) {
    const taskId = event.task_id ?? event.payload?.task_id;
    if (!taskId) {
      return;
    }
    if (!this._tasks.has(taskId)) {
      return this._event("event.rejected", event, {
        error: errorPayload(ErrorCode.TASK_ERROR, `unknown task: ${taskId}`)
      });
    }

    const tracker = this._tasks.get(taskId);

    try {
      const taskEvent = tracker.transition(event.type, event.payload);
      const responses = [this._event("event.acknowledged", event, {
        acknowledged_event_id: event.id
      })];
      responses.push(taskEvent);

      if (tracker.isTerminal()) {
        this._tasks.delete(taskId);
      }

      return responses;
    } catch (err) {
      return this._event("event.rejected", event, {
        error: errorPayload(ErrorCode.TASK_ERROR, err.message)
      });
    }
  }

  _handleSessionOpened(event) {
    if (this._session && this._session.isActive()) {
      return this._event("event.rejected", event, {
        error: errorPayload(ErrorCode.SESSION_ERROR, "session already active", { details: { existing_session: this._session.id } })
      });
    }

    this._session = new AepSession({
      id: event.session_id ?? `sess_${Date.now().toString(36)}`,
      source: this.source,
      version: "0.2"
    });

    this._session.opened();

    return this._session.ready({
      protocol: "harmovela",
      spec_version: "0.2",
      transports: ["stdio"],
      features: ["envelope", "subscription", "task_lifecycle", "error_model"]
    });
  }

  _handleSessionClosed(event) {
    if (!this._session || !this._session.isOpen()) {
      return this._event("event.acknowledged", event, {
        acknowledged_event_id: event.id
      });
    }

    const closedEvent = this._session.close();
    const responses = [this._event("event.acknowledged", event, {
      acknowledged_event_id: event.id
    })];
    if (closedEvent) {
      responses.push(closedEvent);
    }
    return responses;
  }

  startSession(options = {}) {
    this._session = new AepSession(options);
    return this._session.opened();
  }

  _event(type, input, payload) {
    const next = String(++this._sequence).padStart(6, "0");

    return {
      spec_version: input?.spec_version ?? "0.2",
      id: `evt_harness_${next}`,
      type,
      source: this.source,
      target: input?.source,
      topic: input?.topic,
      session_id: input?.session_id,
      conversation_id: input?.conversation_id,
      task_id: input?.task_id,
      correlation_id: input?.correlation_id,
      causation_id: input?.id,
      created_at: this.now(),
      delivery: {
        mode: "best_effort",
        sequence: this._sequence
      },
      payload
    };
  }
}

File: implementations/java/src/main/java/com/axisrobo/aep/Harness.java
package com.axisrobo.aep;

import com.axisrobo.harmovela.event.router.EventRouter;
import com.axisrobo.harmovela.event.session.Session;
import com.axisrobo.harmovela.governance.GovernancePolicy;
import java.time.Instant;
import java.util.*;

public class Harness {

    private static final String SOURCE = "harness:harmovela";
    private int sequence;
    private final Map<String, Map<String, Object>> subscriptions = new LinkedHashMap<>();
    private final Map<String, TaskTracker> tasks = new LinkedHashMap<>();
    private final EventRouter router = new EventRouter();
    private final DeliveryTracker delivery;
    private final List<Map<String, Object>> audit = new ArrayList<>();
    private Session session;

    public Harness() {
        this.delivery = new DeliveryTracker();
        setupDeliveryRouter();
        router
            .on(e -> "capabilities.requested".equals(e.get("type")), this::handleCapabilities)
            .on(e -> "subscription.requested".equals(e.get("type")), this::handleSubscriptionRequested)
            .on(e -> "subscription.cancelled".equals(e.get("type")), this::handleSubscriptionCancelled)
            .on(e -> "task.submitted".equals(e.get("type")), this::handleTaskSubmitted)
            .on(e -> {
                var t = (String) e.get("type");
                return t != null && t.startsWith("task.") && !"task.submitted".equals(t);
            }, this::handleTaskEvent)
            .on(e -> "session.opened".equals(e.get("type")), this::handleSessionOpened)
            .on(e -> "session.closed".equals(e.get("type")), this::handleSessionClosed)
            .on(e -> "session.error".equals(e.get("type")), this::handleSessionError);
    }

    public Session getSession() { return session; }
    public Map<String, Map<String, Object>> getSubscriptions() { return subscriptions; }
    public Map<String, TaskTracker> getTasks() { return tasks; }
    public DeliveryTracker getDelivery() { return delivery; }
    public List<Map<String, Object>> getAudit() { return audit; }

    private void setupDeliveryRouter() {
        router
            .on(e -> "event.acknowledged".equals(e.get("type")), this::handleInboundAck)
            .on(e -> "event.redelivered".equals(e.get("type")), this::handleInboundRedeliver)
            .on(e -> "event.dead_lettered".equals(e.get("type")), this::handleInboundDeadLetter);
    }

    public List<Map<String, Object>> handle(Map<String, Object> value) {
        var errs = com.axisrobo.aep.Envelope.validate(value);
        if (!errs.isEmpty()) {
            return List.of(newEvent("event.rejected", value, Map.of(
                "errors", errs, "error", Errors.errorPayload(Errors.INVALID_ENVELOPE, errs.get(0), false)
            )));
        }

        var type = (String) value.get("type");
        if (!com.axisrobo.aep.EventTypes.isStandardEventType(type) && (type == null || !type.startsWith("session."))) {
            return List.of(newEvent("event.rejected", value, Map.of(
                "errors", List.of("type not in standard draft registry: " + type),
                "error", Errors.errorPayload(Errors.INVALID_EVENT_TYPE, "unknown event type: " + type, false)
            )));
        }

        if (!"0.2".equals(value.get("spec_version"))) {
            return List.of(newEvent("event.rejected", value, Map.of(
                "errors", List.of("unsupported protocol version: " + value.get("spec_version")),
                "error", Errors.errorPayload(Errors.UNSUPPORTED_VERSION, "unsupported version " + value.get("spec_version"), false)
            )));
        }

        var actorId = (String) value.get("actor_id");
        var requestedAction = (String) value.get("requested_action");
        if (actorId != null && requestedAction != null) {
            var tenantId = (String) value.get("tenant_id");
            var targetTenantId = (String) value.get("target_tenant_id");
            var correlationId = (String) value.get("correlation_id");
            var causationId = (String) value.get("causation_id");
            @SuppressWarnings("unchecked")
            var roles = (List<String>) value.getOrDefault("roles", List.of());

            var decision = GovernancePolicy.authorize(actorId, tenantId, roles, requestedAction, targetTenantId);
            var auditEntry = new LinkedHashMap<String, Object>();
            auditEntry.put("actor_id", actorId);
            auditEntry.put("tenant_id", tenantId);
            auditEntry.put("action", requestedAction);
            auditEntry.put("target_tenant_id", targetTenantId);
            auditEntry.put("allowed", decision.allowed());
            auditEntry.put("correlation_id", correlationId);
            auditEntry.put("causation_id", causationId);
            audit.add(auditEntry);

            if (!decision.allowed()) {
                return List.of(newEvent("event.rejected", value, Map.of(
                    "error", Errors.errorPayload(Errors.UNAUTHORIZED, "governance denied: " + decision.reason(), false)
                )));
            }
        }

        var deliveryMeta = value.get("delivery");
        if (deliveryMeta instanceof Map<?,?> dm && dm.get("mode") instanceof String mode && !mode.isEmpty()) {
            var eventId = (String) value.get("id");
            var sessionId = (String) value.getOrDefault("session_id", "_default");
            delivery.track(eventId, sessionId);
        }

        var routed = router.dispatch(value);
        if (!routed.isEmpty()) return routed;

        return List.of(newEvent("event.acknowledged", value, Map.of("acknowledged_event_id", value.get("id"))));
    }

    private Object handleCapabilities(Map<String, Object> event) {
        return newEvent("capabilities.declared", event, Map.of(
            "protocol", "harmovela", "spec_version", "0.2",
            "transports", List.of("stdio"),
            "delivery_modes", List.of("best_effort", "at_least_once", "replayable"),
            "features", List.of("envelope_validation", "event_type_registry", "subscription_matching",
                "session_lifecycle", "task_lifecycle", "error_model", "event_routing")
        ));
    }

    @SuppressWarnings("unchecked")
    private Object handleSubscriptionRequested(Map<String, Object> event) {
        var payload = (Map<String, Object>) event.getOrDefault("payload", Map.of());
        var subId = "sub_" + String.format("%04d", nextSeq());

        var hasFilter = payload.containsKey("types") || payload.containsKey("source")
            || payload.containsKey("target") || payload.containsKey("topic");
        if (!hasFilter) {
            return newEvent("subscription.rejected", event, Map.of(
                "subscription_id", subId, "filter", payload,
                "error", Errors.errorPayload(Errors.SUBSCRIPTION_REJECTED, "subscription must include at least one filter criterion", false)
            ));
        }

        subscriptions.put(subId, Map.of("id", subId, "filter", payload, "created_at", Instant.now().toString()));
        return newEvent("subscription.created", event, Map.of("subscription_id", subId, "filter", payload));
    }

    @SuppressWarnings("unchecked")
    private Object handleSubscriptionCancelled(Map<String, Object> event) {
        var payload = (Map<String, Object>) event.get("payload");
        if (payload != null && payload.get("subscription_id") instanceof String subId) {
            subscriptions.remove(subId);
        }
        return newEvent("event.acknowledged", event, Map.of("acknowledged_event_id", event.get("id")));
    }

    @SuppressWarnings("unchecked")
    private Object handleTaskSubmitted(Map<String, Object> event) {
        var taskId = (String) event.getOrDefault("task_id", null);
        if (taskId == null && event.get("payload") instanceof Map<?, ?> p) {
            taskId = (String) p.get("task_id");
        }
        if (taskId == null) taskId = "task_" + System.currentTimeMillis();

        if (tasks.containsKey(taskId)) {
            return newEvent("event.rejected", event, Map.of(
                "error", Errors.errorPayload(Errors.TASK_ERROR, "duplicate task id: " + taskId, false)
            ));
        }

        var source = (String) event.getOrDefault("source", "unknown");
        var tracker = new TaskTracker(taskId, source);
        tracker.accept();
        tasks.put(taskId, tracker);

        return newEvent("task.accepted", event, Map.of("task_id", taskId, "status", "accepted"));
    }

    @SuppressWarnings("unchecked")
    private Object handleTaskEvent(Map<String, Object> event) {
        var taskId = (String) event.getOrDefault("task_id", null);
        if (taskId == null && event.get("payload") instanceof Map<?, ?> p) {
            taskId = (String) p.get("task_id");
        }
        if (taskId == null) {
            return newEvent("event.rejected", event, Map.of(
                "error", Errors.errorPayload(Errors.TASK_ERROR, "task event missing task_id", false)
            ));
        }

        var tracker = tasks.get(taskId);
        if (tracker == null) {
            return newEvent("event.rejected", event, Map.of(
                "error", Errors.errorPayload(Errors.TASK_ERROR, "unknown task: " + taskId, false)
            ));
        }

        var eventType = (String) event.get("type");
        var payload = event.get("payload") instanceof Map<?, ?> p ? (Map<String, Object>) p : null;
        var taskEvent = tracker.transition(eventType, payload);
        if (taskEvent == null) {
            return newEvent("event.rejected", event, Map.of(
                "error", Errors.errorPayload(Errors.TASK_ERROR, "illegal task transition: " + tracker.getState() + " for task " + taskId, false)
            ));
        }

        var responses = new ArrayList<>(List.of(newEvent("event.acknowledged", event, Map.of("acknowledged_event_id", event.get("id")))));
        responses.add(taskEvent);

        if (tracker.isTerminal()) tasks.remove(taskId);
        return responses;
    }

    private Object handleSessionOpened(Map<String, Object> event) {
        if (session != null && session.isActive()) {
            return newEvent("event.rejected", event, Map.of(
                "error", Errors.errorPayload(Errors.SESSION_ERROR, "session already active", false)
            ));
        }
        var sessionId = (String) event.getOrDefault("session_id", null);
        session = new Session(sessionId, SOURCE, "0.2");
        var opened = session.opened();

        var ready = Map.<String, Object>of(
            "spec_version", "0.2",
            "id", "evt_sess_ready_" + System.currentTimeMillis(),
            "type", "session.ready",
            "source", SOURCE,
            "session_id", session.getId(),
            "created_at", Instant.now().toString(),
            "payload", Map.of("session_id", session.getId(),
                "capabilities", Map.of("protocol", "harmovela", "spec_version", "0.2",
                    "transports", List.of("stdio"),
                    "features", List.of("envelope", "subscription", "task_lifecycle", "error_model")))
        );

        return List.of(opened, ready);
    }

    private Object handleSessionClosed(Map<String, Object> event) {
        var responses = new ArrayList<>(List.of(newEvent("event.acknowledged", event, Map.of("acknowledged_event_id", event.get("id")))));
        if (session != null && session.isOpen()) {
            var closed = session.close();
            if (closed != null) responses.add(closed);
        }
        return responses;
    }

    private Object handleSessionError(Map<String, Object> event) {
        return newEvent("event.acknowledged", event, Map.of("acknowledged_event_id", event.get("id")));
    }

    @SuppressWarnings("unchecked")
    private Object handleInboundAck(Map<String, Object> event) {
        var payload = (Map<String, Object>) event.get("payload");
        if (payload != null) {
            var eventId = (String) payload.get("acknowledged_event_id");
            if (eventId != null) delivery.ack(eventId);
        }
        return newEvent("event.acknowledged", event, Map.of("acknowledged_event_id", event.get("id")));
    }

    @SuppressWarnings("unchecked")
    private Object handleInboundRedeliver(Map<String, Object> event) {
        var payload = (Map<String, Object>) event.get("payload");
        if (payload != null) {
            var eventId = (String) payload.get("original_event_id");
            if (eventId != null) delivery.nack(eventId);
        }
        return newEvent("event.acknowledged", event, Map.of("acknowledged_event_id", event.get("id")));
    }

    @SuppressWarnings("unchecked")
    private Object handleInboundDeadLetter(Map<String, Object> event) {
        var payload = (Map<String, Object>) event.get("payload");
        if (payload != null) {
            var eventId = (String) payload.get("original_event_id");
            if (eventId != null) {
                var reason = new HashMap<String, Object>();
                reason.put("code", "unknown");
                if (payload.get("error") instanceof Map<?,?> err) {
                    reason.putAll((Map<String, Object>) err);
                }
                delivery.deadLetter(eventId, reason);
            }
        }
        return newEvent("event.acknowledged", event, Map.of("acknowledged_event_id", event.get("id")));
    }

    private int nextSeq() { return ++sequence; }

    private Map<String, Object> newEvent(String type, Map<String, Object> input, Map<String, Object> payload) {
        var seq = nextSeq();
        var map = new LinkedHashMap<String, Object>();
        map.put("spec_version", input.getOrDefault("spec_version", "0.2"));
        map.put("id", "evt_harness_" + String.format("%06d", seq));
        map.put("type", type);
        map.put("source", SOURCE);
        putIfNotNull(map, "target", input.get("source"));
        putIfNotNull(map, "session_id", input.get("session_id"));
        putIfNotNull(map, "task_id", input.get("task_id"));
        putIfNotNull(map, "causation_id", input.get("id"));
        map.put("created_at", Instant.now().toString());
        map.put("delivery", Map.of("mode", "best_effort", "sequence", seq));
        map.put("payload", payload);
        return map;
    }

    private static void putIfNotNull(Map<String, Object> map, String key, Object value) {
        if (value != null) {
            map.put(key, value);
        }
    }
}
